### ISSUE TITLE
SDK Unable to access TwilioVideoSDK.Error from Swift

### ISSUE THREAD

**Issue by mmimeault** [twilio/twilio-video-ios#103]
### Description

I am trying to access the enum of Error defined in TVIError.h from swift.
I tried the exact same way as shown here: https://github.com/twilio/video-quickstart-ios/blob/master/VideoCallKitQuickStart/ViewController.swift#L391-L393

I am using Podfile to setup Twilio using the latest version.

Ideally I would like to switch/case on the error. For now I will write manually the error code.

#### Code

```swift
if (error as NSError).code != TwilioVideoSDK.Error.roomRoomCompletedError.rawValue {
    reason = .failed
}
```

#### Work around
For now I will write manually the error code Integer.

#### Expected Behavior

Can access the lib error code

#### Actual Behavior

Doesn't compile

#### Logs
When compiling, I get this error:
```
Type 'TwilioVideoSDK' has no member 'Error'
```

### Versions

All relevant version information for the issue.

#### Video iOS SDK

 TwilioVideo (3.2.5) from PodFile

#### Xcode

11.3.1

#### iOS Version

13.3.1

#### iOS Device

iPhone 6S

**Comment by paynerc** [comment]
@mmimeault,

Are you having this issue in your own app or with the Quickstart? I just tested this again in the Quickstart and it is compiling properly Xcode 11.5. There has to be something different in your project that is causing this issue.

I am also creating an internal ticket to see if there is something better we can do with the Error definitions in the SDK to make them more Swift friendly. However this may require waiting until our next major release as it would most likely be a breaking API change.

Ryan

**Comment by mmimeault** [comment]
Hello,
thanks for the support. This is happening in our app.
I could try with the quickstart and compare with our setup.

But I followed the guide to install the sdk (with podfile).

Thanks

**Comment by paynerc** [comment]
@mmimeault.

Have you had any luck resolving this issue?

I have been doing some testing and it looks like we should be able to define `TVIError` using the `NS_ERROR_ENUM` macro and not cause a breaking change with Semantic Versioning. This approach makes consuming `TVIError` objects in Swift more Swift idiomatic.

For instance, in the VideoCallKitQuickStart we currently have the following code:

```.swift
func roomDidDisconnect(room: Room, error: Error?) {
    logMessage(messageText: "Disconnected from room \(room.name), error = \(String(describing: error))")

    if !self.userInitiatedDisconnect, let uuid = room.uuid, let error = error {
        var reason = CXCallEndedReason.remoteEnded
   
        if (error as NSError).code != TwilioVideoSDK.Error.roomRoomCompletedError.rawValue {
            reason = .failed
        }

        self.callKitProvider.reportCall(with: uuid, endedAt: nil, reason: reason)
    }
...
```

When I define `TVIError` with the `NS_ERROR_ENUM` macro, the code can be changed to the following:

```.swift
func roomDidDisconnect(room: Room, error: Error?) {
    logMessage(messageText: "Disconnected from room \(room.name), error = \(String(describing: error))")

    if !self.userInitiatedDisconnect, let uuid = room.uuid, let error = error as? TwilioVideoSDK.Error {
        var reason = CXCallEndedReason.remoteEnded

        if error.code != .roomRoomCompletedError {
            reason = .failed
        }

        self.callKitProvider.reportCall(with: uuid, endedAt: nil, reason: reason)
    }
...
```

This approach then lets us `switch`/`case` statements on the `error.code`. I am working on getting this into the next `TwilioVideo` release and will update this ticket when it has been shipped.

Thank you,

Ryan

**Comment by mmimeault** [comment]
Hello,

Thank you for the follow up.

I haven't spent time to compare my setup with the sample app.
Unfortunately, I still have to use the "hardcoded" error code.
It still give me  `Type 'TwilioVideoSDK' has no member 'Error'` error message when compiling.

Do I required a specific swift version for it to work?

Thanks

**Comment by paynerc** [comment]
@mmimeault,

I just released Twilio Video iOS `3.6.0` which contains the `NS_ERROR_ENUM` approach which should give you better error interoperability in Swift.

As far as Swift versions, our Quickstart applications target Swift 4.2 and the [Video App](https://github.com/twilio/twilio-video-app-ios) targets Swift 5. What version of Swift are you using in your projects?

Ryan

**Comment by paynerc** [comment]
@mmimeault,

It's been a while since we released `3.6.0` with the `NS_ERROR_ENUM` support and am going to close this issue. Please feel free to reopen this issue if you are still experiencing issues.

Ryan
